src/lib/FlightTasks/tasks/FlightTaskManualAltitude.cpp

```diff
@@ -179,15 +216,11 @@ void FlightTaskManualAltitude::_respectMaxAltitude()
 {
 	if (PX4_ISFINITE(_dist_to_bottom)) {
 
-		// check if there is a valid minimum distance to ground
-		const float min_distance_to_ground = (PX4_ISFINITE(_constraints.min_distance_to_ground)) ?
-						     _constraints.min_distance_to_ground : 0.0f;
-
-		// if there is a valid maximum distance to ground, gradually decrease speed limit upwards from
-		// minimum distance to maximum distance
+		// if there is a valid maximum distance to ground, linearly increase speed limit with distance
+		// below the maximum, preserving control loop vertical position error gain.
 		if (PX4_ISFINITE(_constraints.max_distance_to_ground)) {
-			_constraints.speed_up = math::gradual(_dist_to_bottom, min_distance_to_ground, _constraints.max_distance_to_ground,
-							      _max_speed_up, 0.0f);
+			_constraints.speed_up = math::constrain(MPC_Z_P.get() * (_constraints.max_distance_to_ground - _dist_to_bottom),
```

Comment:
I personally prefer to avoid core-position control parameters like MPC_Z_P. Why cannot we not just limit the speed upwards linearly from max velocity to 0 at maximum altitude?

Reply:
Because that is a de-facto position error gain than can either be too sluggish  if the distance between min alt and max alt is large or can result in overshoot if too small. By using the gain we guarantee the fastest ascent possible without excessive overshoot.